shared_objects/lib2.c

```diff
@@ -0,0 +1,61 @@
+/**
+ * A more complicated example of sealed entries displaying how they can be used to protect private
+ * data that will and can be given to the user of the library as a pointer. In this case Car_priv is
+ * the private fields and methods of the object that is given to the user, and it is protected by
+ * exploiting the properties of the sealed entries (The fact that the capability can only be jumped
+ * to not modified in any way).
+ */
+
+#include <cheri/cheri.h>
+#include <cheri/cheric.h>
+#include <sys/mman.h>
+
+#include "../include/instructions.h"
+#include "../include/regs.h"
+#include "include/lib2.h"
+
+static struct Car *arena;
+static uint32_t size;
+
+void init() {}
+
+void crash()
+{
+	printf("CRASH!\n");
+}
+
+void honk(struct Car_priv *priv, struct Car *car)
+{
+	printf("%p %p\n", priv, car);
+	printf("HONK!\n");
+	if (car->speed > priv->maxSpeed)
+	{
+		priv->crash();
+	}
+}
+
+struct Car *new_car()
+{
+	uint32_t *mem =
+		mmap(NULL, 4096, PROT_READ | PROT_WRITE | PROT_EXEC, MAP_ANON | MAP_PRIVATE, -1, 0);
+
+	uint32_t idx = 0;
+	mem[idx++] = auipcc(cs2, 0);
+	mem[idx++] = clc_128(cs3, cs2, 32);
+	mem[idx++] = cincoffsetimm(ca0, cs2, 32 + 16);
+	mem[idx++] = cincoffsetimm(ca1, cs2, 32 + 16 + sizeof(struct Car_priv));
+	mem[idx++] = cjalr(cnull, cs3);
+	void **ptrs = mem;
+	ptrs[2] = honk;
+
+	uint32_t functions_size = 32 + 16;
```

Comment:
Is there any providence we can make explicit for the magic literals: `32` and `16` here and above ?

Reply:
Not really, it is the size of the instructions + the size of the functions pointers. I will add a comment about that.